compiler/qsc_qasm3/src/parse.rs

```diff
@@ -0,0 +1,265 @@
+        let include_errors = self
+            .source
+            .includes()
+            .iter()
+            .flat_map(QasmSourceTrait::all_errors)
+            .map(|e| self.map_error(e));
+
+        self_errors.extend(include_errors);
+        self_errors
+    }
+
+    #[must_use]
+    pub fn errors(&self) -> Vec<WithSource<crate::Error>> {
+        self.source
+            .errors()
+            .iter()
+            .map(|e| self.map_error(e.clone()))
+            .collect()
+    }
+
+    fn map_error(&self, error: crate::Error) -> WithSource<crate::Error> {
+        let path = self.source.path().display().to_string();
+        let source = self.source_map.find_by_name(&path);
+        let offset = source.map_or(0, |source| source.offset);
+
+        let offset_error = error.with_offset(offset);
+
+        WithSource::from_map(&self.source_map, offset_error)
+    }
+}
+
+/// Parse a QASM file and return the parse result.
+/// This function will resolve includes using the provided resolver.
+/// If an include file cannot be resolved, an error will be returned.
+/// If a file is included recursively, a stack overflow occurs.
+pub fn parse_source<S, P, R>(source: S, path: P, resolver: &R) -> miette::Result<QasmParseResult>
+where
+    S: AsRef<str>,
+    P: AsRef<Path>,
+    R: SourceResolver,
+{
+    let res = parse_qasm_source(source, path, resolver)?;
+    Ok(QasmParseResult::new(res))
+}
+
+/// Creates a Q# source map from a QASM parse output. The `QasmSource`
+/// has all of the recursive includes resolved with their own source
+/// and parse results.
+fn create_source_map(source: &QasmSource) -> SourceMap {
+    let mut files: Vec<(Arc<str>, Arc<str>)> = Vec::new();
+    files.push((
+        Arc::from(source.path().to_string_lossy().to_string()),
+        Arc::from(source.source()),
+    ));
+    // Collect all source files from the includes, this
+    // begins the recursive process of collecting all source files.
+    for include in source.includes() {
+        collect_source_files(include, &mut files);
+    }
+    // Map the main source file to the entry point expression
+    // This may be incorrect, but it's the best we can do for now.
+    SourceMap::new(files, Some(Arc::from(source.source())))
+}
+
+/// Recursively collect all source files from the includes
+fn collect_source_files(source: &QasmSource, files: &mut Vec<(Arc<str>, Arc<str>)>) {
+    files.push((
+        Arc::from(source.path().to_string_lossy().to_string()),
+        Arc::from(source.source()),
+    ));
+    for include in source.includes() {
+        collect_source_files(include, files);
+    }
+}
+
+/// Represents a QASM source file that has been parsed.
+#[derive(Clone, Debug)]
+pub struct QasmSource {
+    /// The path to the source file. This is used for error reporting.
+    /// This path is just a name, it does not have to exist on disk.
+    path: PathBuf,
+    /// The source code of the file.
+    source: String,
```

Comment:
Would it be valuable to make this an `Arc<str>` to mimic our own `Source` type? Or even reuse that type?